factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/ZipStep.kt

```diff
@@ -0,0 +1,109 @@
+ *
+ */
+
+package io.qalipsis.core.factory.steps.zip
+
+import io.qalipsis.api.context.StepContext
+import io.qalipsis.api.context.StepName
+import io.qalipsis.api.context.StepStartStopContext
+import io.qalipsis.api.logging.LoggerHelper.logger
+import io.qalipsis.api.steps.AbstractStep
+import io.qalipsis.api.sync.Latch
+import io.qalipsis.core.exceptions.NotInitializedStepException
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.Channel
+
+
+/**
+ * Step to join flows from several [Step]s into a single one. While this steps technically has several
+ * parents, only one is considered at the primary, from which will the current one will inherit the context.
+ *
+ * This step uses the same mechanism than [io.qalipsis.core.factory.steps.singleton.SingletonProxyStep] to receive
+ * data from secondary steps.
+ *
+ * The step execution is suspended until values from all parent steps are provided.
+ *
+ * @author Polina Bril
+ */
+internal class ZipStep<I, O>(
+    id: StepName,
+    private val coroutineScope: CoroutineScope,
+    /**
+     * Configuration for the consumption from right steps.
+     */
+    private val rightSources: Collection<RightSources<out Any>>,
+    /**
+     * Statement to convert the list of values into the output.
+     */
+    private val outputSupplier: (I, Map<StepName, Any?>) -> O
+
+) : AbstractStep<I, O>(id, null) {
+
+    private val consumptionJobs = mutableListOf<Job>()
+
+    private val channel = Channel<Any?>()
+
+    private var running = false
+
+    override suspend fun start(context: StepStartStopContext) {
+        // Starts the coroutines to buffer the data coming from the right steps.
+        val stepName = this.name
+        rightSources.forEach { source ->
+            consumptionJobs.add(
+                coroutineScope.launch {
+                    log.debug { "Starting the coroutine buffering right records from step ${source.sourceStepName}" }
+                    val subscription = source.topic.subscribe(stepName)
+                    while (subscription.isActive()) {
+                        val record = subscription.pollValue()
+                        log.trace { "Adding right record to channel" }
+                        coroutineScope.launch { channel.send(mapOf(record.stepName to record.value)) }
+                    }
+                    log.debug { "Leaving the coroutine buffering right records" }
+                }
+            )
+        }
+        running = true
+    }
+
+    override suspend fun stop(context: StepStartStopContext) {
+        running = false
+        consumptionJobs.forEach { it.cancel() }
+        rightSources.forEach { it.topic.close() }
+        channel.cancel()
+    }
+
+    override suspend fun execute(context: StepContext<I, O>) {
+        if (!running) throw NotInitializedStepException()
+
+        val input = context.receive()
+        val latch = Latch(true, "zip-step-${name}")
+        coroutineScope.launch {
```

Comment:
```suggestion
```
You already are in a suspended scope, no need to start another coroutine in the present case.